src/core/mount.c

```diff
@@ -1750,9 +1783,54 @@ static int mount_load_proc_self_mountinfo(Manager *m, bool set_flags) {
                 if (!device || !path)
                         continue;
 
+                if (ignore_mount(path, fstype))
+                        continue;
+
+                size = snprintf(mount_entry, mount_entry_size, "%s %s %s %s", device, path, fstype, options);
+
+                if ((size_t)size >= mount_entry_size) {
+                        if (mount_entry)
+                                free(mount_entry);
+                        mount_entry_size = 1024 * (size / 1024 + 1);
+                        mount_entry = new(char, mount_entry_size);
+                        if (!mount_entry) {
+                                mount_entry_size = 0;
+                                return -ENOMEM;
+                        }
+                        size = snprintf(mount_entry, mount_entry_size, "%s %s %s %s", device, path, fstype, options);
+                }
+
+                mu = hashmap_get2(mountinfo_hashmap, mount_entry, (void **)&key);
+                if (mu) {
+                        assert(mu->mountinfo_key == key);
+                        if (mu->state == MOUNT_MOUNTED) {
+                                mu->proc_flags |= MOUNT_PROC_IS_MOUNTED;
+                                log_debug("Mount entry '%s' already known, won't set device nor update mount unit", mount_entry);
+                                continue;
+                        }
+                }
+
+                log_debug("Setting up mount entry '%s'", mount_entry);
+
                 device_found_node(m, device, DEVICE_FOUND_MOUNT, DEVICE_FOUND_MOUNT);
 
-                (void) mount_setup_unit(m, device, path, options, fstype, set_flags);
+                (void) mount_setup_unit(m, device, path, options, fstype, set_flags, &mu);
+
+                if (mu->mountinfo_key) {
+                        (void) hashmap_remove(mountinfo_hashmap, mu->mountinfo_key);
+                        mu->mountinfo_key = mfree(mu->mountinfo_key);
+                }
+
+                key = strdup(mount_entry);
+                if (!key)
+                        return -ENOMEM;
+
+                r = hashmap_put(mountinfo_hashmap, key, mu);
```

Comment:
I'll need to have a look at the hashmap code and think about this for a while.

Being able to check equality in a simple comparison (of addresses is good) but that fact is hidden a little bit by the way it's done. Maybe it's simply my lack of familiarity with the systemd hashmap implementation so don't concern yourself with it if you think it's the right way to do it.

Reply:
Well I spent a lot of my sunday on this.
Here the `key` is what is stored in the Mount unit and the *real key* of the hash.
All this works because we use `hashmap_get2()` to compare the **mount_entry** with the final **key**, which will end up having a string comparison.